### PR TITLE
Add a korean tokenizer

### PR DESCRIPTION
**Pull request by databyjp**
### What's being changed:

- Add a Korean tokenizer using the [kagome](https://github.com/ikawaha/kagome) package.
    - Tokenizer instantiation occurs on instantiation if `ENABLE_TOKENIZER_KOREAN` env var is set to "true"
    - Uses the MeCab-ko dictionary (https://github.com/ikawaha/kagome-dict-ko) similar to what SpaCy uses https://spacy.io/usage/models
- Add `ENABLE_TOKENIZER_GSE` env var for enabling GSE (for Chinese)
    -  Kept `USE_GSE` for backwards compatibility
- Refactor `helpers.Tokenizations` so GSE & Kagome tokenizers are only added if the correct env var is set
- Add a few unit tests
- Minor refactors to related tokenizer code & tests

Tested with the following Python code. Note: editable install of the Python client used, as it required `KAGOME_KR = "kagome_kr"` to be added to `weaviate/collections/classes/config.py` to prevent it throwing a validation error on search.

```python
import weaviate
import os
from weaviate.classes.config import Configure, DataType, Property
from weaviate.classes.query import MetadataQuery
import requests

client = weaviate.connect_to_local(
    headers={
        "X-Cohere-Api-Key": os.environ["COHERE_API_KEY"],
    }
)

client.collections.delete("TextChunk")

url = "http://localhost:8080/v1/schema"

payload = {
    "class": "TextChunk",
    "properties": [
        {
            "name": "chunk",
            "dataType": ["text"],
            "tokenization": "kagome_kr"
        },
        {
            "name": "chunk_number",
            "dataType": ["int"],
        },
    ],
    "vectorizer": "text2vec-cohere",
    "moduleConfig": {
        "generative-cohere": {}
    }
}

response = requests.post(url, json=payload)

temp_coll = client.collections.get("TextChunk")

src_text = """
거북선 또는 귀선(龜船, Geobukseon, turtle ship)은 조선 시대의 군함이다. 거북선은 판옥선을 기본으로 하여 판옥선의 갑판 위 외형 전체에 뚜껑을 씌운 뒤 나무판[1]으로 덮은 배다. 주로 나무판이 아니라 철갑이라고 생각하는 사람이 많지만 비용적인 문제나 바닷물에 녹슬기 쉬운 것을 감안할 경우 나무판이라는 것이 훨씬 설득력 있고, 임진왜란 당시 일본배는 매우 얇아 삼나무 목재선을 사용하여도 전투에 문제가 없었다. 또한 이 나무판에는 적병이 못 뛰어오르도록 무수한 송곳과 칼을 꽂았었다. 선수부에는 용머리 모양의 충각 겸 포문을 만들어 그 곳에서 전면부로 화포를 쏘게 했고 선미부에는 거북이 꼬리를 세우고 역시 화포를 쏘았다. 결국 거북선은 완전 무장으로 승조원을 보호한 채 안전한 곳에서 앞뒤와 선체 측면의 포문으로 전후좌우 각각 6개씩 화포를 발사할 수 있었다.[2]

조선 수군의 지휘관 이순신이 임진왜란 직전에 건조하여 임진왜란 중 사천 해전에서 첫 출전한 이래 칠천량 해전에서 패배하기 전까지 일본 수군과의 16전에서 16승을 하는데 크게 기여하여 일본 수군의 공포의 대명사가 되었다. 조선왕조실록 등에서는 거북선을 한자로 귀선(龜船)으로 표기하고 있다.[3] 임진왜란 이후 일본인들에게는 샤치호코(鯱, 상상의 동물)와 닮은 보쿠카이센 혹은 깃카이센, 기카이센(亀甲船)으로 불렸다는 설이 있다.[4] 1597년 음력 7월 16일 새벽 칠천량 해전에서 일본군에 의해 모두 침몰되었다. 임진왜란 이후에도 만들어졌으나, 임진왜란 당시와 비교해서 모양과 크기가 조금씩 변형되었다.

1973년 9월 대한민국에서는 500원권 지폐의 앞면에 이순신과 거북선을, 뒷면에는 현충사를 도안으로 만들어 쓰기도 하였다. 1966년 이후로 발행된 5원 동전 앞면에도 거북선이 도안되었다.[5] 옥포대첩에는 나가지 않았다.
"""

chunk_len = 50
chunks = [src_text[i*chunk_len:(i+1)*chunk_len] for i in range(len(src_text)//chunk_len+1)]

with temp_coll.batch.dynamic() as batch:
    for i, chunk in enumerate(chunks):
        batch.add_object(
            properties={
                "chunk": chunk,
                "chunk_number": i,
            }
        )

for q in ["거북선", "거북", "귀선", "배", "일본", "수군", "일본수군"]:
    response = temp_coll.query.bm25(
        query=q,
        return_metadata=MetadataQuery(score=True, explain_score=True),
        limit=3
    )
    print("\n" + "=" * 50)
    print(f"Query: {q}")
    for o in response.objects:
        print(o.properties["chunk"])
        print(f"ID: {o.properties['chunk_number']}, Score: {o.metadata.score}, Explain: {o.metadata.explain_score}")

client.close()
```

Outputs:

```text
==================================================
Query: 거북선

거북선 또는 귀선(龜船, Geobukseon, turtle ship)은 조선 시대의 군함
ID: 0, Score: 0.5433447957038879, Explain: , BM25F_거북선_frequency:1, BM25F_거북선_propLength:20
이다. 거북선은 판옥선을 기본으로 하여 판옥선의 갑판 위 외형 전체에 뚜껑을 씌운 뒤 나무
ID: 1, Score: 0.5334352254867554, Explain: , BM25F_거북선_frequency:1, BM25F_거북선_propLength:21
 역시 화포를 쏘았다. 결국 거북선은 완전 무장으로 승조원을 보호한 채 안전한 곳에서 앞뒤
ID: 7, Score: 0.523880660533905, Explain: , BM25F_거북선_frequency:1, BM25F_거북선_propLength:22

==================================================
Query: 거북
각 겸 포문을 만들어 그 곳에서 전면부로 화포를 쏘게 했고 선미부에는 거북이 꼬리를 세우고
ID: 6, Score: 1.126643419265747, Explain: , BM25F_거북_frequency:1, BM25F_거북_propLength:26

==================================================
Query: 귀선

거북선 또는 귀선(龜船, Geobukseon, turtle ship)은 조선 시대의 군함
ID: 0, Score: 1.0052704811096191, Explain: , BM25F_귀선_frequency:1, BM25F_귀선_propLength:20
 수군의 공포의 대명사가 되었다. 조선왕조실록 등에서는 거북선을 한자로 귀선(龜船)으로 표
ID: 11, Score: 0.9357381463050842, Explain: , BM25F_귀선_frequency:1, BM25F_귀선_propLength:24

==================================================
Query: 배
판[1]으로 덮은 배다. 주로 나무판이 아니라 철갑이라고 생각하는 사람이 많지만 비용적인 
ID: 2, Score: 0.9044584631919861, Explain: , BM25F_배_frequency:1, BM25F_배_propLength:26
란 당시 일본배는 매우 얇아 삼나무 목재선을 사용하여도 전투에 문제가 없었다. 또한 이 나
ID: 4, Score: 0.8895899653434753, Explain: , BM25F_배_frequency:1, BM25F_배_propLength:27

==================================================
Query: 일본
천량 해전에서 패배하기 전까지 일본 수군과의 16전에서 16승을 하는데 크게 기여하여 일본
ID: 10, Score: 1.121975064277649, Explain: , BM25F_일본_frequency:2, BM25F_일본_propLength:21
 16일 새벽 칠천량 해전에서 일본군에 의해 모두 침몰되었다. 임진왜란 이후에도 만들어졌으
ID: 14, Score: 0.7843273878097534, Explain: , BM25F_일본_frequency:1, BM25F_일본_propLength:24
란 당시 일본배는 매우 얇아 삼나무 목재선을 사용하여도 전투에 문제가 없었다. 또한 이 나
ID: 4, Score: 0.7456463575363159, Explain: , BM25F_일본_frequency:1, BM25F_일본_propLength:27

==================================================
Query: 수군
천량 해전에서 패배하기 전까지 일본 수군과의 16전에서 16승을 하는데 크게 기여하여 일본
ID: 10, Score: 0.8272411823272705, Explain: , BM25F_수군_frequency:1, BM25F_수군_propLength:21
 수군의 공포의 대명사가 되었다. 조선왕조실록 등에서는 거북선을 한자로 귀선(龜船)으로 표
ID: 11, Score: 0.7843273878097534, Explain: , BM25F_수군_frequency:1, BM25F_수군_propLength:24
와 선체 측면의 포문으로 전후좌우 각각 6개씩 화포를 발사할 수 있었다.[2]

조선 수군
ID: 8, Score: 0.7581090331077576, Explain: , BM25F_수군_frequency:1, BM25F_수군_propLength:26

==================================================
Query: 일본수군
천량 해전에서 패배하기 전까지 일본 수군과의 16전에서 16승을 하는데 크게 기여하여 일본
ID: 10, Score: 1.9492162466049194, Explain: , BM25F_일본_frequency:2, BM25F_일본_propLength:21, BM25F_수군_frequency:1, BM25F_수군_propLength:21
 16일 새벽 칠천량 해전에서 일본군에 의해 모두 침몰되었다. 임진왜란 이후에도 만들어졌으
ID: 14, Score: 0.7843273878097534, Explain: , BM25F_일본_frequency:1, BM25F_일본_propLength:24
 수군의 공포의 대명사가 되었다. 조선왕조실록 등에서는 거북선을 한자로 귀선(龜船)으로 표
ID: 11, Score: 0.7843273878097534, Explain: , BM25F_수군_propLength:24, BM25F_수군_frequency:1
```

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

File: adapters/repos/db/helpers/tokenizer.go [hunk 1]
```diff
@@ -150,18 +170,67 @@ func tokenizeGSE(in string) []string {
...
+type KagomeTokenizers struct {
+	Korean *kagomeTokenizer.Tokenizer
+}
+
+var (
+	tokenizers     KagomeTokenizers
+	kagomeInitLock sync.Mutex
+)
+
+func InitializeKagomeTokenizerKr() error {
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer.go:188]: this method can be a private one

File: adapters/repos/db/helpers/tokenizer.go [hunk 2]
```diff
@@ -150,18 +170,67 @@ func tokenizeGSE(in string) []string {
...
+	// Acquire lock to prevent initialization race
+	kagomeInitLock.Lock()
+	defer kagomeInitLock.Unlock()
+
+	if os.Getenv("ENABLE_KOREAN_TOKENIZER") == "true" {
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer.go:193]: Can we just enable this tokenizer by default?

Comment (antas-marcin) [reply adapters/repos/db/helpers/tokenizer.go:193]: oh I can see that we do this check also from GSE tokenizer, hmm, maybe let's keep this

File: adapters/repos/db/helpers/tokenizer.go [hunk 3]
```diff
@@ -150,18 +170,67 @@ func tokenizeGSE(in string) []string {
...
+		dictInstance := koDict.Dict()
+		tokenizer, err := kagomeTokenizer.New(dictInstance)
+		if err != nil {
+			log.Printf("failed to create Korean tokenizer: %v", err)
+			return err
+		}
+
+		tokenizers.Korean = tokenizer
+		KagomeKrEnabled = true
+		log.Printf("successfully created Korean tokenizer")
```
[the remark this answers is not in the document]
Comment (databyjp) [reply adapters/repos/db/helpers/tokenizer.go]: Have added a new logrus logger instance - not sure if I was supposed to somehow inherit the global one and log with it?

File: adapters/repos/db/helpers/tokenizer.go (hunk 1 above)
Comment (databyjp) [reply adapters/repos/db/helpers/tokenizer.go]: Ah oops, thanks!

File: adapters/repos/db/helpers/tokenizer.go [hunk 4]
```diff
@@ -150,18 +175,63 @@ func tokenizeGSE(in string) []string {
...
+		if err != nil {
+			return err
+		}
+
+		tokenizers.Korean = tokenizer
+		KagomeKrEnabled = true
+		return nil
+	} else {
+		return nil
+	}
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer.go:214]: you can get rid of else and just return nil after if {}

File: adapters/repos/db/helpers/tokenizer_test.go [hunk 5]
```diff
@@ -127,21 +147,27 @@ func TestTokenize(t *testing.T) {
...
 func TestTokenizeAndCountDuplicates(t *testing.T) {
-	input := "Hello You Beautiful World! hello you beautiful world!"
-
 	type testCase struct {
+		input        string
 		tokenization string
 		expected     map[string]int
 	}
 
+	_ = os.Setenv("ENABLE_TOKENIZER_KAGOME_KR", "true")
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer_test.go:156]: in tests you can use `t.Setenv` method

File: adapters/repos/db/helpers/tokenizer.go [hunk 6]
```diff
@@ -18,30 +18,41 @@ import (
...
+	if os.Getenv("ENABLE_TOKENIZER_KAGOME_KR") == "true" {
+		Tokenizations = append(Tokenizations, models.PropertyTokenizationKagomeKr)
+	}
 	init_gse()
+	_ = initializeKagomeTokenizerKr()
 }
 
 func init_gse() {
-	if os.Getenv("USE_GSE") == "true" {
+	if os.Getenv("USE_GSE") == "true" || os.Getenv("ENABLE_TOKENIZER_GSE") == "true" {
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer.go:55]: to check if value is enabled you can use our helper method [configbase.Enabled](https://github.com/weaviate/weaviate/blob/main/usecases/configbase/config_base.go#L16) then you can use it like [here](https://github.com/weaviate/weaviate/blob/main/usecases/config/environment.go#L41) in this example

File: adapters/repos/db/helpers/tokenizer_test.go [hunk 7]
```diff
@@ -63,6 +64,25 @@ func TestTokenise(t *testing.T) {
 
 	tokens = Tokenize(models.PropertyTokenizationGse, "The quick brown fox jumps over the lazy dog")
 	assert.Equal(t, []string{"t", "h", "e", "q", "u", "i", "c", "k", "b", "r", "o", "w", "n", "f", "o", "x", "j", "u", "m", "p", "s", "o", "v", "e", "r", "t", "h", "e", "l", "a", "z", "y", "d", "o", "g", "the", "quick", "brown", "fox", "jumps", "over", "the", "lazy", "dog"}, tokens)
+
+	// Kagome tokenizer for Korean
+	_ = os.Setenv("ENABLE_TOKENIZER_KAGOME_KR", "true")
```
Comment (antas-marcin) [inline adapters/repos/db/helpers/tokenizer_test.go:69]: here also you can use `t.Setenv` method

File: usecases/schema/class.go [hunk 8]
```diff
@@ -581,8 +581,21 @@ func (h *Handler) validatePropertyTokenization(tokenization string, propertyData
...
-				models.PropertyTokenizationWhitespace, models.PropertyTokenizationLowercase, models.PropertyTokenizationTrigram, models.PropertyTokenizationGse:
+				models.PropertyTokenizationWhitespace, models.PropertyTokenizationLowercase,
+				models.PropertyTokenizationTrigram:
 				return nil
+			case models.PropertyTokenizationGse:
+				if os.Getenv("USE_GSE") == "true" || os.Getenv("ENABLE_TOKENIZER_GSE") == "true" {
+					return fmt.Errorf("the GSE tokenizer is not enabled; set 'ENABLE_TOKENIZER_GSE' to 'true' to enable")
+				} else {
+					return nil
+				}
```
Comment (antas-marcin) [inline usecases/schema/class.go:592]: in Golang we tend to omit `else` statements if possible, so in this case you can just `return nil` without `else`

File: usecases/schema/class.go [hunk 9]
```diff
@@ -581,8 +581,21 @@ func (h *Handler) validatePropertyTokenization(tokenization string, propertyData
...
+			case models.PropertyTokenizationKagomeKr:
+				if os.Getenv("ENABLE_TOKENIZER_KAGOME_KR") != "true" {
+					return fmt.Errorf("the Korean tokenizer is not enabled; set 'ENABLE_TOKENIZER_KAGOME_KR' to 'true' to enable")
+				} else {
+					return nil
+				}
```
Comment (antas-marcin) [inline usecases/schema/class.go:598]: same here you can just do `return nil` without the `else`

**Review by antas-marcin — APPROVED** [review]

**Pull request merged**